tfsdk/plan.go

```diff
@@ -163,6 +172,105 @@ func (p *Plan) SetAttribute(ctx context.Context, path *tftypes.AttributePath, va
 	return diags
 }
 
+// pathExists walks the current state and returns true if the path can be reached.
+// The value at the path may be null or unknown.
+func (p Plan) pathExists(ctx context.Context, path *tftypes.AttributePath) (bool, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	_, remaining, err := tftypes.WalkAttributePath(p.Raw, path)
+
+	if err != nil {
+		if errors.Is(err, tftypes.ErrInvalidStep) {
+			return false, diags
+		}
+
+		diags.AddAttributeError(
+			path,
+			"Plan Read Error",
+			"An unexpected error was encountered trying to read an attribute from the plan. This is always an error in the provider. Please report the following to the provider developer:\n\n"+
+				fmt.Sprintf("Cannot walk attribute path in plan: %s", err),
+		)
+		return false, diags
+	}
+
+	return len(remaining.Steps()) == 0, diags
+}
+
+// setAttributeTransformFunc recursively creates a value based on the current
+// Plan values along the path. If the value at the path does not yet exist,
+// this will perform recursion to add the child value to a parent value,
+// creating the parent value if necessary.
+func (p Plan) setAttributeTransformFunc(ctx context.Context, path *tftypes.AttributePath, tfVal tftypes.Value, diags diag.Diagnostics) (transformFunc, diag.Diagnostics) {
```

Comment:
Same here, is this the same implementation between plan and state?

Reply:
See https://github.com/hashicorp/terraform-plugin-framework/pull/165#discussion_r714147857